File: modules/screen-time/ios/ScreenTimeModule.swift
import ExpoModulesCore
import FamilyControls
import ScreenTimeCore

/// JavaScript entry point. It only translates between JS values and
/// `ScreenTimeEngine`; every rule lives in ScreenTimeCore, shared with the extensions.
public final class ScreenTimeModule: Module {
  private static var engine: ScreenTimeEngine { .live }

  public func definition() -> ModuleDefinition {
    Name("ScreenTime")

    Function("getStatus") { () -> ScreenTimeStatus in
      Self.status(engine: Self.engine)
    }

    AsyncFunction("requestAuthorization") { () async throws -> ScreenTimeStatus in
      do {
        try await AuthorizationCenter.shared.requestAuthorization(for: .individual)
      } catch {
        throw AuthorizationFailedException(error.localizedDescription)
      }
      // Apply any lock that couldn't take effect before access was granted.
      Self.engine.reconcile()
      return Self.status(engine: Self.engine)
    }

    AsyncFunction("presentFlaggedAppsPicker") { (maxApps: Int?, promise: Promise) in
      guard Self.engine.canEditFlaggedApps() else {
        throw FlaggedAppsLockedException()
      }
      guard let presenter = appContext?.utilities?.currentViewController() else {
        throw NoPresenterException()
      }
      // `runOnQueue(.main)` below guarantees we're on the main thread here.
      MainActor.assumeIsolated {
        FlaggedAppsPickerSheet.present(
          from: presenter,
          selection: FlaggedApps.shared.selection(),
          maxApps: maxApps
        ) { chosen in
          if let chosen {
            FlaggedApps.shared.save(chosen)
            Self.engine.reconcile()
          }
          promise.resolve(Self.status(engine: Self.engine))
        }
      }
    }
    .runOnQueue(.main)

    AsyncFunction("startSession") { (id: Int, modeName: String, endsAtMs: Double) in
      Self.engine.startSession(id: id, modeName: modeName, endsAt: Self.date(fromMs: endsAtMs))
    }

    AsyncFunction("pauseSession") {
      Self.engine.pauseSession()
    }

    AsyncFunction("resumeSession") { (endsAtMs: Double) in
      Self.engine.resumeSession(endsAt: Self.date(fromMs: endsAtMs))
    }

    AsyncFunction("endSession") { (id: Int) -> [Double] in
      Self.engine.endSession(id: id).map { $0.occurredAt.timeIntervalSince1970 * 1000 }
    }

    AsyncFunction("lockDown") { (untilMs: Double) in
      Self.engine.lockDown(until: Self.date(fromMs: untilMs))
    }

    AsyncFunction("reconcile") {
      Self.engine.reconcile()
    }

    View(FlaggedAppsView.self) {
      Prop("revision") { (view: FlaggedAppsView, _: Int) in
        view.reload()
      }
    }
  }

  private static func status(engine: ScreenTimeEngine) -> ScreenTimeStatus {
    let status = ScreenTimeStatus()
    status.authorization = authorizationName(AuthorizationCenter.shared.authorizationStatus)
    status.flaggedAppCount = FlaggedApps.shared.selection().applicationTokens.count
    status.canEditFlaggedApps = engine.canEditFlaggedApps()
    return status
  }

  private static func authorizationName(_ status: AuthorizationStatus) -> String {
    switch status {
    case .approved:
      return "approved"
    case .denied:
      return "denied"
    case .notDetermined:
      return "notDetermined"
    default:
      // iOS 26.4 added `approvedWithDataAccess`; it grants everything we use.
      if #available(iOS 26.4, *), status == .approvedWithDataAccess {
        return "approved"
      }
      return "notDetermined"
    }
  }

  private static func date(fromMs milliseconds: Double) -> Date {
    Date(timeIntervalSince1970: milliseconds / 1000)
  }
}

struct ScreenTimeStatus: Record {
  @Field var authorization: String = "notDetermined"
  @Field var flaggedAppCount: Int = 0
  @Field var canEditFlaggedApps: Bool = true
}

final class AuthorizationFailedException: GenericException<String>, @unchecked Sendable {
  override var reason: String {
    "Screen Time access wasn't granted: \(param)"
  }
}

final class FlaggedAppsLockedException: Exception, @unchecked Sendable {
  override var reason: String {
    "Flagged apps can't be changed during a focus session or lockdown"
  }
}

final class NoPresenterException: Exception, @unchecked Sendable {
  override var reason: String {
    "There's no screen to present the app picker from"
  }
}

File: modules/screen-time/src/ScreenTimeModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

import type { ScreenTimeStatus } from './ScreenTime.types';

declare class ScreenTimeNativeModule extends NativeModule {
  getStatus(): ScreenTimeStatus;
  requestAuthorization(): Promise<ScreenTimeStatus>;
  presentFlaggedAppsPicker(maxApps: number | null): Promise<ScreenTimeStatus>;
  startSession(id: number, modeName: string, endsAtMs: number): Promise<void>;
  pauseSession(): Promise<void>;
  resumeSession(endsAtMs: number): Promise<void>;
  endSession(id: number): Promise<number[]>;
  lockDown(untilMs: number): Promise<void>;
  reconcile(): Promise<void>;
}

// iOS only. Elsewhere (and in builds without the module) every call is a no-op.
const native = requireOptionalNativeModule<ScreenTimeNativeModule>('ScreenTime');

export const ScreenTime = {
  isAvailable: native !== null,

  getStatus(): ScreenTimeStatus | null {
    return native?.getStatus() ?? null;
  },

  async requestAuthorization(): Promise<ScreenTimeStatus | null> {
    return native ? native.requestAuthorization() : null;
  },

  /** Opens Apple's app picker. `maxApps` is the free-plan cap; `null` means unlimited. */
  async presentFlaggedAppsPicker(maxApps: number | null): Promise<ScreenTimeStatus | null> {
    return native ? native.presentFlaggedAppsPicker(maxApps) : null;
  },

  async startSession(id: number, modeName: string, endsAt: Date): Promise<void> {
    await native?.startSession(id, modeName, endsAt.getTime());
  },

  async pauseSession(): Promise<void> {
    await native?.pauseSession();
  },

  async resumeSession(endsAt: Date): Promise<void> {
    await native?.resumeSession(endsAt.getTime());
  },

  /** Ends the session natively and returns when each "Open anyway" distraction happened. */
  async endSession(id: number): Promise<Date[]> {
    const timestamps = (await native?.endSession(id)) ?? [];
    return timestamps.map(ms => new Date(ms));
  },

  async lockDown(until: Date): Promise<void> {
    await native?.lockDown(until.getTime());
  },

  /** Brings the shields in line with the current time. Call when the app becomes active. */
  async reconcile(): Promise<void> {
    await native?.reconcile();
  },
};
